system/lib/wasmfs/proxied_async_js_impl_backend.h

```diff
@@ -0,0 +1,192 @@
+                                    async_callback_t callback,
+                                    CppCallbackState* state);
+void _wasmfs_jsimpl_async_write(js_index_t backend,
+                                js_index_t index,
+                                const uint8_t* buffer,
+                                size_t length,
+                                off_t offset,
+                                async_callback_t callback,
+                                CppCallbackState* state);
+void _wasmfs_jsimpl_async_read(js_index_t backend,
+                               js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset,
+                               async_callback_t callback,
+                               CppCallbackState* state);
+void _wasmfs_jsimpl_async_get_size(js_index_t backend,
+                                   js_index_t index,
+                                   async_callback_t callback,
+                                   CppCallbackState* state);
+}
+
+namespace wasmfs {
+
+class ProxiedAsyncJSImplFile : public DataFile {
+  emscripten::SyncToAsync& proxy;
+
+  js_index_t getBackendIndex() {
+    static_assert(sizeof(backend_t) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(getBackend());
+  }
+
+  js_index_t getFileIndex() {
+    static_assert(sizeof(this) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(this);
+  }
+
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_write(
+        getBackendIndex(),
+        getFileIndex(),
+        buf,
+        len,
+        offset,
+        [](CppCallbackState* state) { (*state->resume)(); },
+        &state);
+    });
+
+    return state.result;
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    // The caller should have already checked that the offset + len does
+    // not exceed the file's size.
+    assert(offset + len <= getSize());
+
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_read(
+        getBackendIndex(),
+        getFileIndex(),
+        buf,
+        len,
+        offset,
+        [](CppCallbackState* state) { (*state->resume)(); },
+        &state);
+    });
+
+    return state.result;
+  }
+
+  void flush() override {}
+
+  size_t getSize() override {
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_get_size(
+        getBackendIndex(),
+        getFileIndex(),
+        [](CppCallbackState* state) { (*state->resume)(); },
+        &state);
+    });
+
+    return state.offset;
```

Comment:
Would it be worth having different callback state types for the different function signatures, or even for each individual method? On the one hand it would be more types to keep track of, but on the other hand there would be no ambiguity about which fields to use.

Reply:
I think that adds complexity and I'm not sure what ambiguity is removed? `offset` is, by its type, the proper field to use when a call returns an offset. I guess if we had a call that returns two offsets we'd have a problem - but I don't think we have such a syscall?

Reply:
Ok, we can keep it in mind as an option if we do run into any ambiguity.